js/ui/status/missionGame.js

```diff
@@ -0,0 +1,528 @@
+    packingBox.add(subMenuItem.actor);
+    packingBox.add(subMenuItem.menu.actor);
+
+    menu._connectSubMenuSignals(subMenuItem, subMenuItem.menu);
+    menu._connectItemSignals(subMenuItem);
+    subMenuItem._closingId = menu.connect('open-state-changed', function(self, open) {
+        if (!open)
+            subMenuItem.menu.close(BoxPointer.PopupAnimation.FADE);
+    });
+}
+
+/**
+ * wrapTextWith
+ *
+ * Naturally wrap lines at constant, appending prefix to each line,
+ * returning an array with the wrapped lines.
+ *
+ * We need to do this instead of using the line-wrap property on
+ * ClutterText since that messes with the text's allocation
+ * and causes the text to be tightly packed, instead of
+ * overflowing like we want it to.
+ */
+function wrapTextWith(text, constant, prefix) {
+    let lines = [prefix];
+    let lineIndex = 0;
+    let lineCounter = prefix.length;
+    let textCounter = 0;
+    let lastSpace = 0;
+
+    while (textCounter < text.length) {
+        textCounter++;
+
+        /* Break a line when we either hit a space,
+         * hit a newline, when the line (without spaces)
+         * is too long or when we're at the end of the text */
+        const breakCondition = (text[textCounter] == ' ' ||
+                                text[textCounter] == '\n' ||
+                                textCounter == text.length ||
+                                textCounter - lastSpace >= constant);
+        if (breakCondition) {
+            /* Hit a space, see if we can append this
+             * word, otherwise start a new line */
+            let append = text.slice(lastSpace, textCounter);
+            if (lineCounter + append.length > constant) {
+                lines.push(prefix);
+                lineIndex++;
+                lineCounter = prefix.length;
+
+                /* Strip leading and trailing whitespace and newlines
+                 * and remove any leading newlines as well. */
+                lines[lineIndex] = lines[lineIndex] + append.trim().replace(/\n/g, "");;
+            } else {
+                /* Just replace newlines, but don't touch whitespace. */
+                lines[lineIndex] = lines[lineIndex] + append.replace(/\n/g, "");;
+            }
+
+            /* If we encountered a '\n', we also need to
+             * break lines, though this will just
+             * involve inserting a new line. */
+            if (text[textCounter] === '\n') {
+                lines.push(prefix);
+                lineIndex++;
+                lineCounter = prefix.length;
+            }
+
+            lineCounter += append.length;
+            lastSpace = textCounter;
+        }
+    }
+
+    return lines;
+}
+
+function copyObject(src) {
+    let dst = {};
+    Lang.copyProperties(src, dst);
+    return dst;
+}
+
+const ScrolledLabel = new Lang.Class({
+    Name: 'ScrolledLabel',
+    Extends: St.Label,
+
+    _init: function(params) {
+        let parentParams = copyObject(params);
```

Comment:
Ah thanks, thats really helpful. copyObject was used here so that we dont pass params to base objects that they dont understand, but if Params.parse can help us deal with that then all the better.